soundmanager/sound asset controller

mathiassand opened the pull request:
The first implementation of a sound manager to give feedback to the user, by grabbing sounds from an array, and creating gameobjects only when needed. The sounds include feedback when the user makes specific interactions: 

1. Hitting green mole
2. Hitting neutral mole
3. Hitting red mole
4. Hitting the board (behind the moles)
5. Moving out of the motor space
6. Clicking outside of the motor space
7. Moving into the motor space again
8. Countdown sound (when pressing start game)

bastianilso reviewed (commented):
Hey, generally looks good, I like the idea of a static class - but only if it doesn't cause us too much hassle. I'm less fan of having two classes. It's also not very readable what is happening inside `SoundGameAssets.cs` for others. 

Here is an idea:
We could declare SoundAudioClip as a struct outside the SoundManager class:
```
[System.Serializable]
public struct SoundAudioClip {
    public string name;
        public AudioClip audio;
}
```

..and then create a static struct array inside the static SoundManager class:
```
    [SerializeField]
    public static SoundAudioClip[] testStruct;
```

This lets us adaptively adjust the size of the array and drag n' drop the sounds into the inspector in Unity:
![image](https://user-images.githubusercontent.com/3967945/96126201-58d38b80-0ef4-11eb-8e5a-d1633f4d48bc.png)

You can still use your `foreach()` loop inside GetAudioClip() to find the right sound then.

mathiassand commented:
Committed a non-static version of the soundmanager. Implemented an Awake() function as well, as we discussed.

![image](https://user-images.githubusercontent.com/25637660/96553045-e3c9d280-12b4-11eb-9721-0a505398d069.png)

bastianilso reviewed (changes requested):
The way you create the SoundManager class now looks good. 👍 

As for PlaySound() hmm...I had some more reflection time on this. In Virtual Reality, most sounds usually come from objects in your environment.  So if we play every sound from a new game object, we loose the "3D" (sound coming from left, from right) etc).

This means that inherently what we want is to make the object's asking the sound manager to play the sound an audio source themselves, or to pass along what target game object produces it. Ie. if we want to play the sound of a "mole hit" - the hit sound should come from the mole. If we want to play the sound of "countdown", the sound should come from the countdown wall.

So how about something like this?
```
public PlaySound(GameObject source, Sound sound) {

   if (source is null) {
    // we assume that it doesnt matter where the sound comes from if source is null, so we play it from sound manager himself.
    this.audiosource.playoneshot(sound);
   } 

   if (source has no audio source) {
   source.Add component<AudioSource>
   }

   audiosource = source. get component < audio source>
   audiosource.play one shot( sound)
}
```

hendrikknoche commented:
sounds good. is sound rendered in 3D? that's great and definitely tie to mole

mathiassand commented:
tried to implement the changes you suggested - not sure if that's what you meant entirely. let me know :)

bastianilso commented on Assets/Scripts/Game/SoundManager.cs:
```diff
@@ -44,11 +44,24 @@ private void Awake()
+        AudioSource _audioSource = source.GetComponent<AudioSource>();
+        _audioSource.PlayOneShot(GetAudioClip(sound));
+
+        //GameObject soundGameObject = new GameObject("Sound"); // create a gameobject called the corresponding sound
+        //AudioSource audioSource = soundGameObject.AddComponent<AudioSource>(); // attach an audiosource to the gameobject
+        //audioSource.PlayOneShot(GetAudioClip(sound)); // pass in the sound that needs to be played - this is taken from the array 
```
outcommented code can be removed.

bastianilso reviewed (changes requested):
Looks good!
OK lets do some cleanup. I have added just some comments for efficiency and some comments for general clean of the code.

Then we need to:
- Squash your commits into 1 commit.
- [Write a proper commit message](https://chris.beams.io/posts/git-commit/).
- Rebase your work to the latest version in master. This can be done by removing changes that conflict with master (ie. your changes to MainScene) and redo those changes. Unfortunately, I don't think Github Desktop allows you to "edit" your commits. Here is a [tutorial on how to do it from the git command-line](https://www.youtube.com/watch?v=ElRzTuYln0M).
- Add in the references to SoundManager, use PlaySound() where it should be used and test. (The VR equipment is on my desk, if your laptop is powerful enough to test it out, otherwise you can wait until I'm back next week Tuesday).

mathiassand commented:
changed some of the lines that you requested - return in the if statement, created the audiosource variable a little earlier, so that we can use it in the if statement (simplified code), and then I deleted some of the comments.
